lib/gems/pending/util/miq_file_storage.rb

```diff
@@ -0,0 +1,203 @@
+# This class is meant to be a abstract interface for defining a file_storage
+# class.
+#
+# The storage class can either be of a type of "object storage", which includes:
+# * protocols like FTP
+# * document storage like s3 and OpenStack's Swift
+#
+# And mountable filesystems like:
+# * NFS
+# * SMB
+#
+# The class is meant to allow a shared interface for working with these
+# different forms of file storage, while maintaining their differences in
+# implementation where necessary.  Connection will be handled separately by the
+# subclasses, but they must conform to the top level interface.
+#
+class MiqFileStorage
+  def self.with_interface_class(opts)
+    klass = fetch_interface_class(opts)
+    block_given? ? yield(klass) : klass
+  end
+
+  def self.fetch_interface_class(opts)
+    return nil unless opts[:uri]
+
+    require 'uri'
+    scheme, _ = URI.split(URI::DEFAULT_PARSER.escape(opts[:uri]))
+    klass = storage_interface_classes[scheme]
+    klass.new_with_opts(opts) if klass
+  end
+  private_class_method :fetch_interface_class
+
+  def self.storage_interface_classes
+    @storage_interface_classes ||= Interface.descendants.each_with_object({}) do |klass, memo|
+      memo[klass.uri_scheme] = klass if klass.uri_scheme
```

Comment:
What happens when a class's `#uri_scheme` is nil? Will we not get an instance? Should we raise an error in `.fetch_interface_class` or here?

Reply:
@carbonin So with this, I don't think the class would be an issue, since this `if` check here basically prevents nils from being part of the hash.

But, I think the question should have been:

> what happens when the `storage_interface_class[sheme]` lookup returns `nil`?

Right now, when it is called from `.fetch_interface_class`, it will bomb on `.new_with_opts` with a `NoMethodError` on `nil`.  Not the best error, I agree.

I was considering just defaulting to `MiqLocalMountSession`.  The reason I considered doing that is when we are back in `ManageIQ/manageiq` land, we could technically refactor the `evm:db:local:*` tasks to just alias to the remote ones, and then URIs that just have a local path will work through the same interface.  But wasn't sure if that was a good choice

I think we could either do that, or just raise a more friendly error.  I am fine with either.

Reply:
One thing I was considering might be an issue with this is when we add classes after the fact.  For example, if we were to move S3 and Swift implementations their respective `manageiq-provider-*` plugins.

Probably not an issue for now, but leaving this hear as a note.

Reply:
Okay, that makes sense but I think I would rather an error when we can't find a class for a particular uri scheme.

Right now it actually looks like you might get `nil` from `.fetch_interface_class`.